pkg/reconciler/celltenant/cell_tenant.go

```diff
@@ -0,0 +1,262 @@
+	if err != nil {
+		logger.Error("Failed to create Pub/Sub client", zap.Error(err))
+		return err
+	}
+	pubsubReconciler := reconcilerutilspubsub.NewReconciler(client, r.Recorder)
+
+	// Check if topic exists, and if not, create it.
+	topicID := b.GetTopicID()
+	topicConfig := &pubsub.TopicConfig{Labels: b.GetLabels()}
+	if r.DataresidencyStore != nil {
+		if r.DataresidencyStore.Load().DataResidencyDefaults.ComputeAllowedPersistenceRegions(topicConfig, r.ClusterRegion) {
+			logger.Debug("Updated Topic Config AllowedPersistenceRegions for Broker", zap.Any("topicConfig", *topicConfig))
+		}
+	}
+
+	topic, err := pubsubReconciler.ReconcileTopic(ctx, topicID, topicConfig, b.Object(), b.StatusUpdater())
+	if err != nil {
+		return err
+	}
+	// TODO(grantr): this isn't actually persisted due to webhook issues.
+	//TODO uncomment when eventing webhook allows this
+	//b.Status.TopicID = topic.ID()
+
+	// Check if PullSub exists, and if not, create it.
+	subID := b.GetSubscriptionName()
+	subConfig := pubsub.SubscriptionConfig{
+		Topic:  topic,
+		Labels: b.GetLabels(),
+		//TODO(grantr): configure these settings?
+		// AckDeadline
+		// RetentionDuration
+	}
+	if _, err := pubsubReconciler.ReconcileSubscription(ctx, subID, subConfig, b.Object(), b.StatusUpdater()); err != nil {
+		return err
+	}
+
+	// TODO(grantr): this isn't actually persisted due to webhook issues.
+	//TODO uncomment when eventing webhook allows this
+	//b.Status.SubscriptionID = sub.ID()
+
+	return nil
+}
+
+func (r *Reconciler) deleteDecouplingTopicAndSubscription(ctx context.Context, b Statusable) error {
+	logger := logging.FromContext(ctx)
+	logger.Debug("Deleting decoupling topic")
+
+	// get ProjectID from metadata if projectID isn't set
+	projectID, err := utils.ProjectIDOrDefault(r.ProjectID)
+	if err != nil {
+		logger.Error("Failed to find project id", zap.Error(err))
+		b.StatusUpdater().MarkTopicUnknown("FinalizeTopicProjectIdNotFound", "Failed to find project id: %v", err)
+		b.StatusUpdater().MarkSubscriptionUnknown("FinalizeSubscriptionProjectIdNotFound", "Failed to find project id: %v", err)
+		return err
+	}
+
+	client, err := r.getClientOrCreateNew(ctx, projectID, b.StatusUpdater())
+	if err != nil {
+		logger.Error("Failed to create Pub/Sub client", zap.Error(err))
+		return err
+	}
+	pubsubReconciler := reconcilerutilspubsub.NewReconciler(client, r.Recorder)
+
+	// Delete topic if it exists. Pull subscriptions continue pulling from the
+	// topic until deleted themselves.
+	topicID := b.GetTopicID()
+	err = multierr.Append(nil, pubsubReconciler.DeleteTopic(ctx, topicID, b.Object(), b.StatusUpdater()))
+	subID := b.GetSubscriptionName()
+	err = multierr.Append(err, pubsubReconciler.DeleteSubscription(ctx, subID, b.Object(), b.StatusUpdater()))
+
+	return err
+}
+
+// CreatePubsubClientFn is a function for pubsub client creation. Changed in testing only.
+var CreatePubsubClientFn reconcilerutilspubsub.CreateFn = pubsub.NewClient
```

Comment:
This var now is exported, not sure if worth the trouble to only export it in testing.